File: Data Structures & Algorithms/max-water-container/submission-0.java
class Solution {
    public int maxArea(int[] heights) {
        int l = heights.length;
        int a = 0;
        int b = l-1;
        int max = area(heights, a, b);
        while (a < b) {
            if (heights[a] < heights[b]) {
                a++;
            } else {
                b--;
            }
            if (area(heights, a, b) > max) {
                max = area(heights, a, b);
            }
        }
        return max;
    }

    private int area(int[] heights, int a, int b) {
        if (heights[a] > heights[b]) {return heights[b] * (b-a);}
        return (heights[a] * (b-a));
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-0.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)
        while (l < r):
            mid = l + (r-l)//2
            hrs = 0
            print(f"l{l}, r{r}")
            for p in piles:
                hrs += (p-1)//mid + 1
            if hrs > h:
                l = mid + 1
            elif hrs <= h:
                r = mid
        return r